File: android/src/main/kotlin/dev/eggnstone/chime/observers/ChimeVideoTileObserver.kt
package dev.eggnstone.chime.observers

import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoTileState
import io.flutter.plugin.common.EventChannel.EventSink
import org.json.JSONObject

class ChimeVideoTileObserver(private val _eventSink: EventSink) : VideoTileObserver
{
    override fun onVideoTileAdded(tileState: VideoTileState)
    {
        val jsonObject = JSONObject()
        jsonObject.put("Name", "OnVideoTileAdded")
        jsonObject.put("Arguments", convertVideoTileStateToJson(tileState))
        _eventSink.success(jsonObject.toString())
    }

    override fun onVideoTilePaused(tileState: VideoTileState)
    {
        val jsonObject = JSONObject()
        jsonObject.put("Name", "OnVideoTilePaused")
        jsonObject.put("Arguments", convertVideoTileStateToJson(tileState))
        _eventSink.success(jsonObject.toString())
    }

    override fun onVideoTileRemoved(tileState: VideoTileState)
    {
        val jsonObject = JSONObject()
        jsonObject.put("Name", "OnVideoTileRemoved")
        jsonObject.put("Arguments", convertVideoTileStateToJson(tileState))
        _eventSink.success(jsonObject.toString())
    }

    override fun onVideoTileResumed(tileState: VideoTileState)
    {
        val jsonObject = JSONObject()
        jsonObject.put("Name", "OnVideoTileResumed")
        jsonObject.put("Arguments", convertVideoTileStateToJson(tileState))
        _eventSink.success(jsonObject.toString())
    }

    override fun onVideoTileSizeChanged(tileState: VideoTileState)
    {
        val jsonObject = JSONObject()
        jsonObject.put("Name", "OnVideoTileSizeChanged")
        jsonObject.put("Arguments", convertVideoTileStateToJson(tileState))
        _eventSink.success(jsonObject.toString())
    }

    private fun convertVideoTileStateToJson(tileState: VideoTileState): JSONObject
    {
        val jsonObject = JSONObject()

        jsonObject.put("AttendeeId", tileState.attendeeId)
        jsonObject.put("IsContent", tileState.isContent)
        jsonObject.put("IsLocalTile", tileState.isLocalTile)
        jsonObject.put("PauseState", tileState.pauseState)
        jsonObject.put("TileId", tileState.tileId)
        jsonObject.put("VideoStreamContentHeight", tileState.videoStreamContentHeight)
        jsonObject.put("VideoStreamContentWidth", tileState.videoStreamContentWidth)

        return jsonObject
    }
}

File: ios/Classes/observers/ChimeVideoTileObserver.swift
//
//  ChimeVideoTileObserver.swift
//  eggnstone_amazon_chime
//
//  Created by Hendrik Lakämper on 30.07.2020
//

import Foundation
import AmazonChimeSDK
import Flutter

public class ChimeVideoTileObserver : VideoTileObserver {
    let _eventSink: FlutterEventSink
    
    init(eventSink: @escaping FlutterEventSink) {
        self._eventSink = eventSink
    }
    
    public func videoTileDidAdd(tileState: VideoTileState) {
        let json = """
        {
        "Name": "OnVideoTileAdded",
        "Arguments": \(self.convertVideoTileStateToJson(tileState: tileState))
        }
        """
        
        _eventSink(json)
    }
    
    public func videoTileDidRemove(tileState: VideoTileState) {
        let json = """
        {
        "Name": "OnVideoTileRemoved",
        "Arguments": \(self.convertVideoTileStateToJson(tileState: tileState))
        }
        """
        
        _eventSink(json)
    }
    
    public func videoTileDidPause(tileState: VideoTileState) {
        let json = """
        {
        "Name": "OnVideoTileDidPause",
        "Arguments": \(self.convertVideoTileStateToJson(tileState: tileState))
        }
        """

        _eventSink(json)
    }
    
    public func videoTileDidResume(tileState: VideoTileState) {
        let json = """
        {
        "Name": "OnVideoTileDidResume",
        "Arguments": \(self.convertVideoTileStateToJson(tileState: tileState))
        }
        """

        _eventSink(json)
    }
    
    public func videoTileSizeDidChange(tileState: VideoTileState) {
        let json = """
        {
        "Name": "OnVideoTileSizeDidChange",
        "Arguments": \(self.convertVideoTileStateToJson(tileState: tileState))
        }
        """

        _eventSink(json)
    }
    
    func convertVideoTileStateToJson(tileState: VideoTileState) -> String {
        return """
        {
        "AttendeeId": "\(tileState.attendeeId)",
        "IsContent": \(tileState.isContent),
        "IsLocalTile": \(tileState.isLocalTile),
        "PauseState": "\(tileState.pauseState)",
        "TileId": \(tileState.tileId),
        "VideoStreamContentHeight": \(tileState.videoStreamContentHeight),
        "VideoStreamContentWidth": \(tileState.videoStreamContentWidth)
        }
        """
    }
}
